woocommerce/woocommerce-ios
# [Mobile Payments] [SDK 2] Ensure "Remove Card" prompt displays when card is left inserted in reader

@ctarda @joshheald - now ready to review

Addresses #4430 

Changes:

StripeCardReaderService
- listens to didReportReaderEvent (added to SDK 2) to listen for chip card insertion and removal events
- keeps track of whether a chip card was inserted
- adds another promise to the chain to wait for inserted card (if any) to be removed

PaymentCaptureOrchestrator
- It used to handle displayMessage and waitingForInput exactly the same - by calling onPresentMessage with the string (which was then ignored by the VM)
- Now it routes displayMessage to a new modal to display the message (the BBPOS is US only and so English strings from the reader can safely be passed to the VM and VC)
- waitingForInput still calls onPresentMessage as before, but without the ignored string (onPresentMessage is used to ask the user to “present” their payment method by tapping, swiping or inserting it)

OrderDetailsViewModel and OrderDetailsViewController
- onClearMessage (to have user remove card) was confusing to me so I replaced it with onDisplayMessage (which will display remove card message or any other message passed to it)

CardPresentModalDisplayMessage
- displays messages from headless card readers, like “Remove Card”

Feedback welcome:
- I'm not 100% happy with the spin lock approach to the promise. Suggestions welcome, but short of using PromiseKit, I'm not sure there are any other alternatives.

Background and Other approaches considered:

- SDK2 added the didReportReaderEvent that turns out to be essential for properly managing chip card state - it tells (for chip cards) when a card is inserted and when it is removed
- But, as currently written, the CardPresentPaymentStore only forwards reader events to its clients when in the midst of collectPaymentMethod. Those events stop when the payment completion fires. Unfortunately, that means that if a chip card is used and not removed before payment completes (as can happen - as the reader will allow payment to complete with a card in it), the clients will not be informed of the removal.
- I considered expanding the availability of those events outside the scope of collectPaymentMethod by adding new actions to CardPresentPaymentStore to allow subscribing to chip card reader events (inserted, removed) and having  PaymentCaptureOrchestrator use those actions to subscribe to those events and keep track of inserted card that needs to be removed
- But the more I worked on that approach, the more I didn't like that the orchestrator and view model were having to know details about how the card reader and chip cards worked.
- So, at the end of it, I went with this approach that keeps the details of headless chip card capture as close to the reader as possible and opens up the door for displaying other messages from headless readers if we want.

Screenshots/Movies:

https://www.youtube.com/watch?v=gk-fcToyFjQ

Note: The reader will send "Remove Card" as a message for display for tap, swipe or insert, but only sends the actual "card inserted" and "card removed" events for insert (chip) flows. That's why you see it for all flows, but it only remains on the screen until the card is removed for insert flows.

Note: When multiple contactless cards were near my reader and I attempted swipe payments, I would occassionally see a "Multiple Contactless Cards" error briefly. Not sure we need to worry about this as it seems unlikely to happen in real life, but I wrote it up https://github.com/woocommerce/woocommerce-ios/issues/5193 - this has probably been happening all along, but only now (with this PR) do we display messages from the reader.

To test:
- Collect a payment with tapping a card. No changes to the flow should be seen.
- Collect a payment with swiping a card. No changes to the flow should be seen.
- Collect a payment with inserting a card. Don't remove it right away. Make sure you are prompted to remove it before processing continues. 

Update release notes:
